### PR TITLE
added for moengage and tested

### PR DESCRIPTION
**Pull request by ruchiramoitra**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ yes ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/153)

### CODE REVIEW

File: integrations/MoEngage/browser.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+/* eslint-disable guard-for-in */
+/* eslint-disable prefer-rest-params */
+/* eslint-disable vars-on-top */
+/* eslint-disable no-var */
+/* eslint-disable no-param-reassign */
+/* eslint-disable func-names */
+import each from "@ndhoule/each";
+import logger from "../../utils/logUtil";
+
+class MoEngage {
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:19]: please include comments involving the destination documentation
any special case handling and why it is required and where it's mentioned in destination docs etc

File: integrations/MoEngage/browser.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+      a.onload = function () {
+        if (n) {
+          i[r] = moe(n);
+        }
+      };
+    })(
+      window,
+      document,
+      "script",
+      "https://cdn.moengage.com/webpush/moe_webSdk.min.latest.js",
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:83]: can we add this with Scriptloader? Also include the check that which protocol the website site is opened with (http/https etc) and depending on that get the destination sdk min file.

File: integrations/MoEngage/browser.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    return !!window.moeBannerText;
+  };
+
+  track(rudderElement) {
+    logger.debug("inside track");
+    // Check if the anonymous id is same as previous session if not a new session will start
+    if (rudderElement.message) {
+      const { anonymousId, event, properties } = rudderElement.message;
+      if (anonymousId) {
+        if (this.initialAnonId !== anonymousId) {
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:118]: also reinitialise  this.initialAnonId to new anonymousID?
also, it's almost expected that anonId will not change, is it better to perform this using userID?

File: integrations/MoEngage/browser.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        }
+      }
+      if (event) {
+        if (properties) {
+          this.moeClient.track_event(event, properties);
+        } else {
+          this.moeClient.track_event(event);
+        }
+      } else {
+        logger.error("Event name not present");
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:129]: perform this and below check earlier to reduce if/else condition?

File: integrations/MoEngage/browser.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+      }
+    } else {
+      logger.error("Payload not correct");
+    }
+  }
+
+  reset() {
+    logger.debug("inside reset");
+    // reset the anonymous id
+    this.initialAnonId = window.rudderanalytics.getAnonymousId();
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:139]: also, it's almost expected that anonId will not change, is it better to perform this using userID?

File: integrations/MoEngage/browser.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    // check if anonymous id is same or not
+    if (this.initialAnonId !== anonymousId) {
+      this.reset();
+    }
+    // if user is present map
+    if (userId) {
+      this.moeClient.add_unique_user_id(userId);
+    }
+    // custom traits mapping context.traits --> moengage properties
+    const traitsMap = {
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:156]: move this const outside of the class?

File: integrations/MoEngage/browser.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+      each(function name(value, key) {
+        // check if name is present
+        if (key === "name") {
+          self.moeClient.add_user_name(value);
+        }
+
+        // if any of the custom properties are present it will be sent
+        // For example : email is present in username then the method will be add_user_name
+
+        if (Object.prototype.hasOwnProperty.call(traitsMap, key)) {
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:177]: can the below 3 loops be merged into one?
perform lookup to traitsMap, if present use **add${key_name}** else use **add_user_attribute**

File: tests/html/script-test.html [hunk 8]
```diff
@@ -33,8 +33,9 @@
         })(method);
       }
 
-      rudderanalytics.load("1cA1w20Fe34TdZrtGshutNY8Fm1", "http://localhost:8080", {
```
Comment (sayan-mitra) [inline tests/html/script-test.html:36]: include this test file inside the moengage dest folder.
also remove so many commented lines and use the final tests that were run.

File: integrations/MoEngage/browser.js [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    logger.debug("inside reset");
+    // reset the anonymous id
+    this.initialAnonId = window.rudderanalytics.getAnonymousId();
+    this.moeClient.destroy_session();
+  }
+
+  identify(rudderElement) {
+    const self = this;
+    const { anonymousId, userId } = rudderElement.message;
+    const { traits } = rudderElement.message.context;
```
Comment (sayan-mitra) [inline integrations/MoEngage/browser.js:146]: put a check if context is present or not.
in general, whenever there is more than one indirection, it's better to check for existence.

**Review by sayan-mitra — CHANGES_REQUESTED** [review]
add code comments heavily.

File: integrations/MoEngage/browser.js (hunk 3 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:118]: changed to user id

File: integrations/MoEngage/browser.js (hunk 2 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:83]: done

File: integrations/MoEngage/browser.js (hunk 1 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:19]: done

File: integrations/MoEngage/browser.js (hunk 5 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:139]: done

File: integrations/MoEngage/browser.js (hunk 6 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:156]: done

File: tests/html/script-test.html (hunk 8 above)
Comment (ruchiramoitra) [reply tests/html/script-test.html:36]: I created a test folder with different cases which I finally ran including the results.

File: integrations/MoEngage/browser.js (hunk 9 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:146]: done

File: integrations/MoEngage/browser.js (hunk 7 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:177]: but we have to loop through traits too right? to send the remaining attributes?

File: integrations/MoEngage/browser.js (hunk 4 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:129]: I did not get this.

File: integrations/MoEngage/browser.js [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+// custom traits mapping context.traits --> moengage properties
+const traitsMap = {
+  firstName: "first_name",
+  lastName: "last_name",
+  firstname: "first_name",
+  lastname: "last_name",
+  email: "email",
+  phone: "mobile",
+  name: "user_name",
+  username: "user_name",
```
Comment (arnab-p) [inline integrations/MoEngage/browser.js:13]: As we're supporting camelCase, should we support `userName` as well?

**Review by arnab-p — CHANGES_REQUESTED** [review]

File: integrations/MoEngage/browser.js (hunk 10 above)
Comment (ruchiramoitra) [reply integrations/MoEngage/browser.js:13]: done

**Review by arnab-p — COMMENTED** [review]
LGTM

**Review by arnab-p — APPROVED** [review]

**Review by sayan-mitra — APPROVED** [review]

**Pull request merged**